node/src/accountant/scanners_utils.rs

```diff
@@ -0,0 +1,540 @@
+        }
+
+        let (biggest, oldest) = all_non_pending_payables
+            .iter()
+            .map(|payable| PayableInfo {
+                balance_wei: payable.balance_wei,
+                age: timestamp
+                    .duration_since(payable.last_paid_timestamp)
+                    .expect("Payable time is corrupt")
+                    .as_secs(),
+            })
+            .fold(
+                Default::default(),
+                |(so_far_biggest, so_far_oldest): (PayableInfo, PayableInfo), payable| {
+                    (
+                        bigger(so_far_biggest, payable),
+                        older(so_far_oldest, payable),
+                    )
+                },
+            );
+
+        format!("Payable scan found {} debts; the biggest is {} owed for {}sec, the oldest is {} owed for {}sec",
+                all_non_pending_payables.len(), biggest.balance_wei, biggest.age,
+                oldest.balance_wei, oldest.age)
+    }
+
+    pub fn separate_errors(
+        sent_payments: &SentPayables,
+        logger: &Logger,
+    ) -> (Vec<Payable>, Vec<BlockchainError>) {
+        sent_payments
+            .payable
+            .iter()
+            .fold((vec![], vec![]), |so_far, payment| {
+                match payment {
+                    Ok(payment_sent) => (plus(so_far.0, payment_sent.clone()), so_far.1),
+                    Err(error) => {
+
+                        logger.warning(|| match &error {
+                            BlockchainError::TransactionFailed { .. } => format!("Encountered transaction error at this end: '{:?}'", error),
+                            x => format!("Outbound transaction failure due to '{:?}'. Please check your blockchain service URL configuration.", x)
+                        });
+
+                        (so_far.0, plus(so_far.1, error.clone()))
+                    }
+                }
+            })
+    }
+
+    pub fn payables_debug_summary(qualified_accounts: &[(PayableAccount, u128)], logger: &Logger) {
+        if qualified_accounts.is_empty() {
+            return;
+        }
+        debug!(logger, "Paying qualified debts:\n{}", {
+            let now = SystemTime::now();
+            qualified_accounts
+                .iter()
+                .map(|(payable, threshold_point)| {
+                    let p_age = now
+                        .duration_since(payable.last_paid_timestamp)
+                        .expect("Payable time is corrupt");
+                    format!(
+                        "{} wei owed for {} sec exceeds threshold: {} wei; creditor: {}",
+                        payable.balance_wei.separate_with_commas(),
+                        p_age.as_secs(),
+                        threshold_point.separate_with_commas(),
+                        payable.wallet
+                    )
+                })
+                .join("\n")
+        })
+    }
+
+    pub trait PayableThresholdsGauge {
```

Comment:
I like traits whenever there are supposed to be multiple implementations. Over here we don't require multiple implementations. I'm not asking you to change it in any way. I prefer to reduce the use of traits and use direct functions or a struct implementation.


It is a personal preference, so I won't force you to think my way, just understanding my perspective is enough.

Reply:
This is a mockable thing. Real and mock.

Reply:
I was aware that there were implementations in both Production and Test Code. The point I was referring to was that I only feel that traits are used in the correct scenarios if:


Scenario 1: It requires multiple implementations in Production code.
Scenario 2: We can't directly test the production code, so we need to create a Mock, hence we'll use traits.

In our case, I felt that it is plausible to say that we can test these functions even if they're direct functions, which can be tested directly. So, just for this case, I felt that it doesn't fall in either of the scenarios.

Now, don't take the other way. I highly respect the efforts that you did over there. I just want to let you know about my way of doing things. In the end, it's up to you to reject some or all of my suggestions. Since, we're both right and the differences lie in the style of coding. In the end, I have huge respect for paying extra efforts for making the codebase stronger.

Feel free to resolve the comment, once you've read it. 😉